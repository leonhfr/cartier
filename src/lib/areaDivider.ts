// Packages.
import * as debug from 'debug';
import * as turf from '@turf/turf';

// Internal.
import { GEOJSON_ZONES_DEFAULT_PRECISION } from '../constants';

// Code.
const debugVerbose = debug('cartier:verbose:areaDivider');

export const areaDivider = (
  area: turf.helpers.Feature<turf.helpers.Polygon>,
  precision: number = GEOJSON_ZONES_DEFAULT_PRECISION
): Array<turf.helpers.Feature<turf.helpers.Polygon>> => {
  debugVerbose(`area (%d): %j`, precision, area);

  const zones: Array<turf.helpers.Feature<turf.helpers.Polygon>> = [];

  // Computing bounding box
  const digits = (precision
    .toString()
    .split('.')
    .pop() as string).length;

  const boundingBox = turf.bbox(area);

  const [minX, minY, maxX, maxY] = boundingBox.map(x =>
    Number((Math.floor(x / precision) * precision).toFixed(digits))
  );

  debugVerbose(`boundingBox: %j`, [minX, minY, maxX, maxY]);

  // Computing zones
  for (let x = minX; x < maxX + precision; x += precision) {
    for (let y = minY; y < maxY + precision; y += precision) {
      const mask = turf.polygon([
        [
          [x, y],
          [x + precision, y],
          [x + precision, y + precision],
          [x, y + precision],
          [x, y],
        ],
      ]);

      const zone = turf.intersect(area, mask);

      debugVerbose(`intersection with mask %j: %j`, mask, zone);

      // No intersection
      if (!zone) {
        continue;
      }

      const type = turf.getType(zone);

      if (type === 'Polygon') {
        zones.push(zone);
      }

      if (type === 'MultiPolygon') {
        const { coordinates } = zone.geometry as turf.helpers.MultiPolygon;
        for (const coords of coordinates) {
          zones.push(turf.polygon(coords));
        }
      }
    }
  }

  debugVerbose(`zones: %j`, zones);

  return zones;
};
