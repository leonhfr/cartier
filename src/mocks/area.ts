// Packages.
import * as turf from '@turf/turf';

// Code.
export const area1: turf.helpers.Feature<turf.helpers.Polygon> = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [2.186450958251953, 41.40346698504464],
        [2.1866226196289062, 41.39123380197929],
        [2.193145751953125, 41.386983841682195],
        [2.2202682495117188, 41.411706989227795],
        [2.1976089477539062, 41.428312571522916],
        [2.2170066833496094, 41.410934533229224],
        [2.198638916015625, 41.42226296732032],
        [2.2140884399414062, 41.409775832009565],
        [2.186450958251953, 41.40346698504464],
      ],
    ],
  },
};

export const area2: turf.helpers.Feature<turf.helpers.Polygon> = {
  type: 'Feature',
  properties: {},
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [2.1508097648620605, 41.41161043273029],
        [2.152419090270996, 41.41225414000288],
        [2.1521401405334473, 41.41281737863403],
        [2.153770923614502, 41.41391165687079],
        [2.154543399810791, 41.41323578131299],
        [2.155337333679199, 41.41482890530818],
        [2.154672145843506, 41.41519901943104],
        [2.1529126167297363, 41.41672772933266],
        [2.150101661682129, 41.416856461155575],
        [2.1487069129943848, 41.41667945483331],
        [2.147698402404785, 41.414651893460444],
        [2.148492336273193, 41.414249592012546],
        [2.149651050567627, 41.414394420820805],
        [2.149994373321533, 41.413638089039786],
        [2.1496081352233887, 41.413139227087825],
        [2.1508312225341797, 41.412415065824376],
        [2.150402069091797, 41.41215758431861],
        [2.1508097648620605, 41.41161043273029],
      ],
    ],
  },
};
