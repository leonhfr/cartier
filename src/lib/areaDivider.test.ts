// Internal.
import { areaDivider } from './areaDivider';
import * as Mocks from '../mocks';

// Code.
describe('areaDivider', () => {
  it('should be defined', () => {
    expect(areaDivider).toBeDefined();
  });
  it('should return the expected output', () => {
    expect(areaDivider(Mocks.area1, 0.01)).toMatchSnapshot();
  });
  it('should return the expected output', () => {
    expect(areaDivider(Mocks.area2)).toMatchSnapshot();
  });
  it('should be safe with the decimals', () => {});
});
