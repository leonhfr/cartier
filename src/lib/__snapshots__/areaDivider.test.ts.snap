// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`areaDivider should return the expected output - default precision 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.15081,
            41.41161,
          ],
          Array [
            2.150402,
            41.412158,
          ],
          Array [
            2.150831,
            41.412415,
          ],
          Array [
            2.149608,
            41.413139,
          ],
          Array [
            2.149994,
            41.413638,
          ],
          Array [
            2.149651,
            41.414394,
          ],
          Array [
            2.148492,
            41.41425,
          ],
          Array [
            2.147698,
            41.414652,
          ],
          Array [
            2.148707,
            41.416679,
          ],
          Array [
            2.150102,
            41.416856,
          ],
          Array [
            2.152913,
            41.416728,
          ],
          Array [
            2.154672,
            41.415199,
          ],
          Array [
            2.155337,
            41.414829,
          ],
          Array [
            2.154543,
            41.413236,
          ],
          Array [
            2.153771,
            41.413912,
          ],
          Array [
            2.15214,
            41.412817,
          ],
          Array [
            2.152419,
            41.412254,
          ],
          Array [
            2.15081,
            41.41161,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;

exports[`areaDivider should return the expected output 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.19,
            41.3890337470489,
          ],
          Array [
            2.1885169722352895,
            41.39,
          ],
          Array [
            2.19,
            41.39,
          ],
          Array [
            2.19,
            41.3890337470489,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.1885169722352895,
            41.39,
          ],
          Array [
            2.186623,
            41.391234,
          ],
          Array [
            2.1864997471593233,
            41.4,
          ],
          Array [
            2.19,
            41.4,
          ],
          Array [
            2.19,
            41.39,
          ],
          Array [
            2.1885169722352895,
            41.39,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.1864997471593233,
            41.4,
          ],
          Array [
            2.186451,
            41.403467,
          ],
          Array [
            2.19,
            41.40427716901255,
          ],
          Array [
            2.19,
            41.4,
          ],
          Array [
            2.1864997471593233,
            41.4,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.1964546580107616,
            41.39,
          ],
          Array [
            2.193146,
            41.386984,
          ],
          Array [
            2.19,
            41.3890337470489,
          ],
          Array [
            2.19,
            41.39,
          ],
          Array [
            2.1964546580107616,
            41.39,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2,
            41.39323174876484,
          ],
          Array [
            2.1964546580107616,
            41.39,
          ],
          Array [
            2.19,
            41.39,
          ],
          Array [
            2.19,
            41.4,
          ],
          Array [
            2.2,
            41.4,
          ],
          Array [
            2.2,
            41.39323174876484,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.19,
            41.40427716901255,
          ],
          Array [
            2.2,
            41.40655997828998,
          ],
          Array [
            2.2,
            41.4,
          ],
          Array [
            2.19,
            41.4,
          ],
          Array [
            2.19,
            41.40427716901255,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2,
            41.42617098546242,
          ],
          Array [
            2.197609,
            41.428313,
          ],
          Array [
            2.2,
            41.426560718522445,
          ],
          Array [
            2.2,
            41.42617098546242,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2,
            41.421162941290696,
          ],
          Array [
            2.198639,
            41.422263,
          ],
          Array [
            2.2,
            41.42142363763066,
          ],
          Array [
            2.2,
            41.421162941290696,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2074250095053185,
            41.4,
          ],
          Array [
            2.2,
            41.39323174876484,
          ],
          Array [
            2.2,
            41.4,
          ],
          Array [
            2.2074250095053185,
            41.4,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.21,
            41.40234722697441,
          ],
          Array [
            2.2074250095053185,
            41.4,
          ],
          Array [
            2.2,
            41.4,
          ],
          Array [
            2.2,
            41.40655997828998,
          ],
          Array [
            2.21,
            41.408842787567394,
          ],
          Array [
            2.21,
            41.40234722697441,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.208952145068047,
            41.42,
          ],
          Array [
            2.21,
            41.41923206324198,
          ],
          Array [
            2.21,
            41.417212329930926,
          ],
          Array [
            2.206888294165037,
            41.42,
          ],
          Array [
            2.208952145068047,
            41.42,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2023083841807902,
            41.42,
          ],
          Array [
            2.21,
            41.415256390243904,
          ],
          Array [
            2.21,
            41.41308021748981,
          ],
          Array [
            2.2014387987507,
            41.42,
          ],
          Array [
            2.2023083841807902,
            41.42,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2,
            41.426560718522445,
          ],
          Array [
            2.208952145068047,
            41.42,
          ],
          Array [
            2.206888294165037,
            41.42,
          ],
          Array [
            2.2,
            41.42617098546242,
          ],
          Array [
            2.2,
            41.426560718522445,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.2,
            41.42142363763066,
          ],
          Array [
            2.2023083841807902,
            41.42,
          ],
          Array [
            2.2014387987507,
            41.42,
          ],
          Array [
            2.2,
            41.421162941290696,
          ],
          Array [
            2.2,
            41.42142363763066,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.218395360999875,
            41.41,
          ],
          Array [
            2.21,
            41.40234722697441,
          ],
          Array [
            2.21,
            41.408842787567394,
          ],
          Array [
            2.214088,
            41.409776,
          ],
          Array [
            2.2138108657003333,
            41.41,
          ],
          Array [
            2.218395360999875,
            41.41,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.22,
            41.411462705183986,
          ],
          Array [
            2.218395360999875,
            41.41,
          ],
          Array [
            2.2138108657003333,
            41.41,
          ],
          Array [
            2.21,
            41.41308021748981,
          ],
          Array [
            2.21,
            41.415256390243904,
          ],
          Array [
            2.217007,
            41.410935,
          ],
          Array [
            2.21,
            41.417212329930926,
          ],
          Array [
            2.21,
            41.41923206324198,
          ],
          Array [
            2.22,
            41.41190340796152,
          ],
          Array [
            2.22,
            41.411462705183986,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.22,
            41.41190340796152,
          ],
          Array [
            2.220268,
            41.411707,
          ],
          Array [
            2.22,
            41.411462705183986,
          ],
          Array [
            2.22,
            41.41190340796152,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
]
`;
